http-mclient/src/Network/HTTP/MClient/Decoder.hs

```diff
@@ -0,0 +1,87 @@
+module Network.HTTP.MClient.Decoder
+  ( Env
+  , ResponseDecoder
+  , ResponseError
+  , SendRequest
+  , decodeJSON
+  , decodeStatus
+  , decodeStatuses
+  , send
+  )
+where
+
+import Control.Arrow (left)
+import Data.Conversions (toText)
+import GHC.Records (HasField(..))
+import MPrelude
+
+import qualified Data.Aeson                 as JSON
+import qualified Data.ByteString            as BS
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import qualified Data.List                  as List
+import qualified Network.HTTP.Client        as HTTP
+import qualified Network.HTTP.Types         as HTTP
+import qualified UnliftIO.Exception         as Exception
+
+type SendRequest = HTTP.Request -> IO (HTTP.Response LBS.ByteString)
+
+type Env env = HasField "httpSendRequest" env SendRequest
+
+data ResponseError
+  = BodyDecodeFailure Text
+  | HTTPError HTTP.HttpExceptionContent
+  | UnexpectedContentType BS.ByteString
+  | MissingContentType
+  | UnexpectedStatusCode HTTP.Status
+  deriving stock Show
+
+instance Exception.Exception ResponseError
+
+type Result a = Either ResponseError a
+
+type ResponseDecoder a = HTTP.Response LBS.ByteString -> Result a
+
+decodeStatus
+  :: HTTP.Status
+  -> ResponseDecoder a
+  -> ResponseDecoder a
+decodeStatus expectedStatus decoder = decodeStatuses [(expectedStatus, decoder)]
+
+decodeStatuses
+  :: [(HTTP.Status, ResponseDecoder a)]
+  -> ResponseDecoder a
+decodeStatuses decoders response
+  = maybe (Left $ UnexpectedStatusCode statusCode) ($ response)
+  $ List.lookup statusCode decoders
+  where
+    statusCode = HTTP.responseStatus response
+
+decodeExpectedContentType :: BS.ByteString -> ResponseDecoder a -> ResponseDecoder a
```

Comment:
There is a reason I opted to NOT use the http-media-types package and its types: Its parsing only works form the serve side perspective, not from client side perspective.

And for an API client use case its enough to simply statically expect a single content type accepting 0 variance / conneg.